api/v1alpha8/types.go

```diff
@@ -291,31 +291,61 @@ type LoadBalancer struct {
 	Tags []string `json:"tags,omitempty"`
 }
 
-// SecurityGroup represents the basic information of the associated
+// SecurityGroupStatus represents the basic information of the associated
 // OpenStack Neutron Security Group.
-type SecurityGroup struct {
-	Name  string              `json:"name"`
-	ID    string              `json:"id"`
-	Rules []SecurityGroupRule `json:"rules"`
+type SecurityGroupStatus struct {
+	Name  string                    `json:"name"`
+	ID    string                    `json:"id"`
+	Rules []SecurityGroupRuleStatus `json:"rules"`
 }
 
-// SecurityGroupRule represent the basic information of the associated OpenStack
+// SecurityGroupRuleSpec represent the basic information of the associated OpenStack
 // Security Group Role.
-type SecurityGroupRule struct {
-	Description     string `json:"description"`
-	ID              string `json:"name"`
-	Direction       string `json:"direction"`
-	EtherType       string `json:"etherType"`
-	SecurityGroupID string `json:"securityGroupID"`
-	PortRangeMin    int    `json:"portRangeMin"`
-	PortRangeMax    int    `json:"portRangeMax"`
-	Protocol        string `json:"protocol"`
-	RemoteGroupID   string `json:"remoteGroupID"`
-	RemoteIPPrefix  string `json:"remoteIPPrefix"`
+// For now this is only used for the allNodesSecurityGroupRules but when we'll
+// other security groups, we'll need to add a validation because
+// Remote* fields are mutually exclusive.
+type SecurityGroupRuleSpec struct {
+	Description         string                     `json:"description,omitempty"`
+	Direction           string                     `json:"direction,omitempty"`
+	EtherType           string                     `json:"etherType,omitempty"`
+	PortRangeMin        int                        `json:"portRangeMin,omitempty"`
+	PortRangeMax        int                        `json:"portRangeMax,omitempty"`
+	Protocol            string                     `json:"protocol,omitempty"`
+	RemoteGroupID       string                     `json:"remoteGroupID,omitempty"`
+	RemoteIPPrefix      string                     `json:"remoteIPPrefix,omitempty"`
+	RemoteManagedGroups []ManagedSecurityGroupName `json:"remoteManagedGroups,omitempty"`
+}
+
+type SecurityGroupRuleStatus struct {
+	ID             string `json:"id"`
+	Description    string `json:"description,omitempty"`
+	Direction      string `json:"direction,omitempty"`
+	EtherType      string `json:"etherType,omitempty"`
+	PortRangeMin   int    `json:"portRangeMin,omitempty"`
+	PortRangeMax   int    `json:"portRangeMax,omitempty"`
+	Protocol       string `json:"protocol,omitempty"`
+	RemoteGroupID  string `json:"remoteGroupID,omitempty"`
+	RemoteIPPrefix string `json:"remoteIPPrefix,omitempty"`
+}
+
+// +kubebuilder:validation:Enum=bastion;controlplane;worker
```

Comment:
add a test for that

Reply:
I think kubebuilder adds this to the CRD and *waves hands*? Can you check the CRD?

Reply:
cool, I'm working on that next.

Reply:
@mdbooth where do you expect tests for this one? in `controllers/suite_test.go` ?

Reply:
The enum is visible in CRD. See https://github.com/kubernetes-sigs/cluster-api-provider-openstack/pull/1826/files#diff-1a379f1e3a8a3141f9200c4629eb4c10e79ef7440e43c2e24287bc36de424d6bR5528-R5536

Enum is just a part of the API spec meaning you can still create an instance of the resource that uses other objects. The filtering is done in Kubernetes during an apply/patch. Sure this could be done in a mock test however I think it would make sense as an e2e test?